<launch>
  <arg name="num_vehicles" default="4" />
  <arg name="include_collision_avoidance" default="false" />

  <group unless="$(arg include_collision_avoidance)">
    <node pkg="carla_multi_vehicle_planner" type="multi_vehicle_spawner.py" name="multi_vehicle_spawner" output="screen">
      <param name="num_vehicles" value="$(arg num_vehicles)" />
      <param name="enable_autopilot" value="false" />
      <param name="vehicle_model" value="vehicle.vehicle.coloredxycar" />
      <param name="target_speed" value="6.0" />
      <!-- Reset CARLA world (and actor ID counter) at startup -->
      <param name="reset_world_on_start" value="true" />
      <!-- Reset CARLA world on shutdown so IDs start from 1 next run -->
      <param name="reset_world_on_shutdown" value="true" />
    </node>

    <node pkg="carla_multi_vehicle_planner" type="networkx_path_planner.py" name="networkx_path_planner" output="screen">
      <env name="PYTHONPATH" value="$(find carla_multi_vehicle_planner)/scripts:$(optenv PYTHONPATH)" />
      <param name="num_vehicles" value="$(arg num_vehicles)" />
      <!-- Smaller map -> denser graph but shorter routes -->
      <param name="graph_resolution" value="0.25" />
      <param name="path_sampling" value="0.2" />
      <param name="min_destination_distance" value="20.0" />
      <param name="max_destination_distance" value="120.0" />
      <param name="override_preempt_min_dist" value="3.0" />
      <param name="destination_retry_limit" value="50" />
    </node>

    <node pkg="carla_multi_vehicle_planner" type="multi_vehicle_controller.py" name="multi_vehicle_controller" output="screen">
      <param name="num_vehicles" value="$(arg num_vehicles)" />
      <!-- Shorter lookahead for smaller map, keep speed moderate -->
      <param name="lookahead_distance" value="2.0" />
      <param name="target_speed" value="6.0" />
    </node>
  </group>

  <include if="$(arg include_collision_avoidance)" file="$(find carla_multi_vehicle_planner)/launch/predictive_collision_avoidance.launch">
    <arg name="num_vehicles" value="$(arg num_vehicles)" />
  </include>

  <node pkg="carla_multi_vehicle_planner" type="bev_visualizer.py" name="bev_visualizer" output="screen">
    <param name="max_vehicle_count" value="6" />
    <param name="update_rate" value="2.0" />
  </node>
</launch>